### PR TITLE
aria2: Rewrite init script

### PR DESCRIPTION
**Pull request by kuoruan**
* Add aria2 user and group.
* Use procd to start service.
* Add more supported options.

Signed-off-by: Hsing-Wang Liao <kuoruan@gmail.com>

Maintainer: me / @kuoruan
Compile tested: LEDE 17.01.2
Run tested: x86_64

### CODE REVIEW

**Comment by kuoruan** [comment]
@Ansuel I have rewrote aria2 and luci-app-aria2, could you please help me test it?

**Comment by Ansuel** [comment]
@kuoruan sure

**Comment by Ansuel** [comment]
@kuoruan 
have this problem
Wed Nov 29 13:32:37 2028 daemon.err aria2c[2943]: See 'aria2c -h'.
Wed Nov 29 13:32:37 2028 daemon.err aria2c[2943]: Usage: aria2c [OPTIONS] [URI | MAGNET | TORRENT_FILE | METALINK_FILE]...
Wed Nov 29 13:32:37 2028 daemon.err aria2c[2943]: Specify at least one URL.

**Comment by kuoruan** [comment]
@Ansuel How did this happened?

**Comment by Ansuel** [comment]
Just flashed new version with old aria2 config

**Comment by kuoruan** [comment]
@Ansuel What's your config file? Can you run aria2 instance?

**Comment by Ansuel** [comment]
@kuoruan ok i removed my list_extra option... think there are some problem with option that are declered 2 times...
also log page doesn't clear if you start and stop aria2 deamon 
could you also add rpc secure and https cert directory to luci app? 
(now is working)

**Comment by kuoruan** [comment]
@Ansuel OK, I will add this options.

**Comment by Ansuel** [comment]
Tell me when you are done so i can test it

**Comment by kuoruan** [comment]
@hnyman Everything done. but check failed.

**Comment by lynxis** [comment]
@kuoruan as quick-and-dirty solution. rebase your PR will trigger the travis again.

**Comment by kuoruan** [comment]
@lynxis still failed.

**Comment by lynxis** [comment]
<code>
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
</code>
Ok, we might want to add `V=s` to the travis check.

**Comment by lynxis** [comment]
@champtar we hit the travis limit in **not** sending anything now.
@kuoruan Since `make check` is succeeding. Treat the result as a "pass"

**Comment by kuoruan** [comment]
@lynxis I use my computer to compile. Only used 6 minutes with i5 single thread. Why it spend so much time with three thread?

**Comment by lynxis** [comment]
@kuoruan it's unknown how many cores the builders have and how many cputime we can use.

**Comment by champtar** [comment]
@lynxis I removed the V=s because I was hitting the 4mb log limit for some packages ...
I will find a workaround, but in the mean time please do ignore the failure

**Comment by kuoruan** [comment]
@champtar Now it's OK, but here seems something wrong with the script? 
https://travis-ci.org/openwrt/packages/builds/264620413#L2284

**Comment by kuoruan** [comment]
@hnyman Travis check pass and test OK, please consider merge.

**Comment by Ansuel** [comment]
Why no merge?

File: net/aria2/Makefile [hunk 1]
```diff
@@ -15,21 +15,27 @@ PKG_SOURCE_URL:=https://github.com/aria2/aria2/releases/download/release-$(PKG_V
...
+	ARIA2_LIBGCRYPT \
+	ARIA2_LIBXML2 \
+	ARIA2_EXPAT \
+	ARIA2_GMP \
+	ARIA2_BITTORRENT \
+	ARIA2_METALINK \
+	ARIA2_SFTP \
+	ARIA2_ASYNC_DNS \
+	ARIA2_COOKIE \
+	ARIA2_WEBSOCKET
```
Comment (yousong) [inline net/aria2/Makefile]: Well, this list needs to be prefixed with `CONFIG_` as they are used to form `.configure_xxx` stamp file.

**Comment by yousong** [comment]
It will be helpful if the list of changes can be part of the commit message body.  The key information I'd like to know is: does the rewrite change some defaults or introduce any incompatibilities?

File: net/aria2/Makefile (hunk 1 above)
Comment (kuoruan) [inline net/aria2/Makefile]: "CONFIG_" will be auto add as prefix, so there is no need it.

**Comment by kuoruan** [comment]
@yousong The only change is "list extra_settings" to "list extra_setting".

File: net/aria2/Makefile (hunk 1 above)
Comment (yousong) [inline net/aria2/Makefile]: Hmm, I cannot find the code doing it in aria2/Makefile, include/package.mk, rules.mk .  Can you shed more light ? ;)

**Comment by yousong** [comment]
I would suggest keeping the old option name to avoid possible troubles of existing users of it.

**Comment by kuoruan** [comment]
@yousong Everything fixed.

File: net/aria2/Makefile (hunk 1 above)
Comment (kuoruan) [reply net/aria2/Makefile]: @yousong 
There are some packages that has custom ```Config.in``` file.
https://github.com/lede-project/source/blob/master/package/libs/openssl/Makefile
https://github.com/lede-project/source/blob/master/package/boot/kexec-tools/Makefile
There ```PKG_CONFIG_DEPENDS``` all has ```CONFIG_``` prefix.
And I have tested adding ```CONFIG_```, works well. So I think we should add ```CONFIG_```.

File: net/aria2/files/aria2.init [hunk 2]
```diff
@@ -1,112 +1,348 @@
...
+		'seed_time:ufloat' \
+		'split:uinteger' \
+		'timeout:uinteger' \
+		'user_agent:string'
+}
 
-	section_enabled "$s" || return 1
+aria2_start() {
+	local section="$1"
+	aria2_validate "$section" || { _err "Validation failed."; return 1; }
```
Comment (yousong) [inline net/aria2/files/aria2.init]: `uci_validate_section` will export those options as global variables and I remembered this causing unexpected results when rewriting shadowsocks-libev: a section inheriting option values set in previous sections.  Will this bug also affect aria2 here?

Comment (kuoruan) [reply net/aria2/files/aria2.init]: 你说的这个问题我也遇到过，但是我已经规避了。
第二个 section 进来的时候会覆盖掉第一个 section 的变量，所以没问题。
Create file: ```/etc/config/test```
```
config test
        option a 'a'
        option b 'b'
        option c 'c'
config test
        option a '1'
```
Create ```/etc/init.d/test```
```
#!/bin/sh /etc/rc.common
USE_PROCD=1

validate_test() {
        uci_validate_section test test "$1" \
                'a:string' \
                'b:string' \
                'c:string'
}

test_func() {
        validate_test "$1"
        echo "a: $a   b: $b    c: $c"
}

start_service() {
        config_load test
        config_foreach test_func test
}
```
run ```/etc/init.d/test start```
output: 
```
a: a   b: b    c: c
a: 1   b:     c:
```

Comment (kuoruan) [reply net/aria2/files/aria2.init]: @yousong 你看看还有问题吗？

Comment (yousong) [reply net/aria2/files/aria2.init]: Okay, looks like the issue won't affect aria2's use case here.  The situation with shadowsocks-libev is more sophisticated that it requires explicit first declaring those option names as `local`.

**Comment by yousong** [comment]
Thanks for the efforts!

File: net/aria2/files/aria2.init [hunk 3]
```diff
@@ -1,112 +1,348 @@
...
+		'check_certificate:or("true","false"):true' \
+		'connect_timeout:uinteger' \
+		'dht_listen_port:string' \
+		'dir:string' \
+		'disable_ipv6:or("true","false")' \
+		'disk_cache:string' \
+		'enable_dht:or("true","false"):true' \
+		'enable_dht6:or("true","false")' \
+		'enable_peer_exchange:("true","false")' \
+		'event_poll:("epoll","kqueue","port","poll","select")' \
```
Comment (yousong) [inline net/aria2/files/aria2.init]: Missing `or` in the above 2 datatype spec?

Comment (kuoruan) [reply net/aria2/files/aria2.init]: @yousong OK. I will make a new PR.

**Pull request merged**